Lib/test/test_free_threading/test_str.py

```diff
@@ -0,0 +1,74 @@
+import sys
+import unittest
+
+from itertools import cycle
+from threading import Thread
+from unittest import TestCase
+
+from test.support import threading_helper
+
+@threading_helper.requires_working_threading()
+class TestStr(TestCase):
+    def test_racing_join_extend(self):
+        '''Test joining a string being extended by another thread'''
+        l = []
+        OBJECT_COUNT = 100_000
+
+        def writer_func():
+            l.extend(map(str, range(OBJECT_COUNT, OBJECT_COUNT*2)))
+
+        def reader_func():
+            while True:
+                count = len(l)
+                ''.join(l)
+                if count == OBJECT_COUNT:
+                    break
+
+        writer = Thread(target=writer_func)
+        readers = []
+        for x in range(30):
```

Comment:
A few suggestions:

* We want the tests to be fast since they are run all the time on the GitHub CI, which is kind of slow and, if I remember correctly, only has ~2 virtual CPUs. We should aim for <100 ms for each of these tests.
* We'll trigger bugs more reliably and quickly by doing repeated smaller amounts of work in the writing thread, rather than one big extend operation.
* Let's use `threading.Event()` for termination.

```suggestion
        ITERS = 100
        READERS = 10
        done_event = threading.Event()

        def writer_func():
            for i in range(ITERS):
                l.extend(map(str, range(i)))
                l.clear()
            done_event.set()

        def reader_func():
            while not done_event.is_set():
                ''.join(l)

        writer = Thread(target=writer_func)
        readers = []
        for x in range(READERS):
```

Reply:
FYI, I chose 30 readers to match the number used in the `test.test_free_threading.test_list`'s `test_racing_iter_append` (the model for this entire test). I'm happy to change it (I did overdo the `OBJECT_COUNT` trying to guarantee a seg fault without the patch), just noting it would make for less consistent style/design.

I'll apply the change manually (there are a couple tweaks needed to make it legal so I can't autocommit this one).

Reply:
Change applied.